aev/src/main/java/com/fingerprintjs/android/aev/AevClientImpl.kt

```diff
@@ -1,48 +1,71 @@
 package com.fingerprintjs.android.aev
 
 
-import com.fingerprintjs.android.fingerprint.Fingerprinter
+import com.fingerprintjs.android.aev.config.ConfigProvider
 import com.fingerprintjs.android.aev.logger.Logger
 import com.fingerprintjs.android.aev.signals.SignalProviderImpl
-import java.util.concurrent.ExecutorService
-import java.util.concurrent.Executors
+import com.fingerprintjs.android.aev.utils.concurrency.callbackToSync
+import com.fingerprintjs.android.aev.utils.concurrency.runInAnotherThread
+import com.fingerprintjs.android.aev.utils.concurrency.runInParallel
+import com.fingerprintjs.android.fingerprint.DeviceIdResult
+import com.fingerprintjs.android.fingerprint.FingerprintResult
+import com.fingerprintjs.android.fingerprint.Fingerprinter
 
 
 internal class AevClientImpl(
     private val ossAgent: Fingerprinter,
     private val apiInteractor: ApiInteractor,
     private val signalProviderBuilder: SignalProviderImpl.SignalProviderBuilder,
-    private val logger: Logger
+    private val logger: Logger,
+    private val configProvider: ConfigProvider,
 ) : AevClient {
 
-    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
-
     override fun getRequestId(listener: (String) -> Unit) {
         getRequestId(listener, {})
     }
 
     override fun getRequestId(listener: (String) -> Unit, errorListener: (String) -> Unit) {
-        executor.execute {
-            logger.debug(this, "Start getting requestId.")
-            ossAgent.getDeviceId { deviceIdResult ->
-                logger.debug(this, "Got deviceId: ${deviceIdResult.deviceId}")
-                ossAgent.getFingerprint { fingerprintResult ->
-                    logger.debug(this, "Got fingerprint: ${fingerprintResult.fingerprint}")
+        runInAnotherThread {
+            runInParallel(
+                {
+                    logger.debug(this, "Start getting requestId.")
+                    callbackToSync<DeviceIdResult> { ossAgent.getDeviceId { emit(it) } }.also {
+                        logger.debug(this, "Got deviceId: ${it.deviceId}")
+                    }
+                },
+                {
+                    logger.debug(this, "Start getting fingerprint.")
+                    callbackToSync<FingerprintResult> { ossAgent.getFingerprint { emit(it) } }.also {
+                        logger.debug(this, "Got fingerprint: ${it.fingerprint}")
+                    }
+                },
+                {
+                    logger.debug(this, "Start getting config.")
+                    configProvider.getConfig().also {
+                        logger.debug(this, "Got config")
+                    }
+                }
+            ).let {
+                val deviceIdResult = it.first.getOrNull()
+                val fingerprintResult = it.second.getOrNull()
+                val config = it.third.getOrNull()
+                if (deviceIdResult != null && fingerprintResult != null && config != null) {
```

Comment:
If something is null, we just stop working, without reporting of error upstairs?

Reply:
Yeah, this doesn't seem right. It worked the same way previously, though. I'll think about what error we should return if something is null.